core-bundle/src/SimpleTokens/Parser.php

```diff
@@ -0,0 +1,231 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\SimpleTokens;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+use Symfony\Component\ExpressionLanguage\ExpressionLanguage;
+
+class Parser
+{
+    /**
+     * @var LoggerInterface|null
+     */
+    private $logger;
+
+    /**
+     * @var ExpressionLanguage
+     */
+    private $expressionLanguage;
+
+    public function __construct(LoggerInterface $logger = null, ExpressionLanguage $expressionLanguage = null)
+    {
+        if (null === $expressionLanguage) {
```

Comment:
Looks strange to me. I think I would make it a required constructor attribute because it cannot work without any `ExpressionLanguage` instance.
I would extract the code that creates a "secure" EL instance so it can be reused by someone else, if requried. Or make it a factory method at least. Wdyt?

Reply:
:+1: for making it a required argument.

I'm not sure about the 'secure' bit though - does it ever make sense to have the `constant` function not disabled/overwritten? I have the feeling people would just forget to disable it when injecting their custom version. Or everyone would have to cater for it if there are multiple decorators in place. :thinking:

Reply:
Yeah well...that's an argument that's basically true for everything. You could also just swap our authenticators for custom ones that do wrong checks and then the whole login setup is insecure :D 

You wanted to register your own functions, right? Then we need to register that EL as a service anyway and give you the option to tag your services. Afaik there's even something like an `ExpressionFunctionProviderInterface`.

Reply:
Yes, we might want to inject a list of tagged providers instead and register them with `$expressionLanguage->registerProvider();`. Otherwise the only extension-friendly way of extending the EL is decorating the simple token parser.